timemachine/fe/mcgregor.py

```diff
@@ -286,6 +292,31 @@ def atom_map_pop(map_1_to_2, map_2_to_1, idx, jdx):
     map_2_to_1[jdx] = UNMAPPED
 
 
+def _graph_is_disconnected(g, mapped_nodes, demapped_nodes, unvisited_nodes):
+    mapped_nodes = set(mapped_nodes)
+    demapped_nodes = set(demapped_nodes)
+    unvisited_nodes = set(unvisited_nodes)
+
+    # check intersections
+    assert mapped_nodes.union(demapped_nodes).union(unvisited_nodes) == set(range(g.n_vertices))
+    assert len(mapped_nodes.intersection(demapped_nodes)) == 0
+    assert len(mapped_nodes.intersection(unvisited_nodes)) == 0
+    assert len(demapped_nodes.intersection(unvisited_nodes)) == 0
```

Comment:
nit: assume we'll want to remove/disable these assertions once bugs are worked out? (this seems like relatively "hot" code)

Reply:
will remove - i will do another performance optimization pass. The code will probably get pretty hairy so I wanted a review on the simplest implementation first.